### PR TITLE
LLM: support llama2 8k input with w4a16.

### PR DESCRIPTION
**Pull request by lalalapotter**
## Description

This PR is to support llama2 8k input with w4a16

- Perf test results:
```shell
# 8k input with split tensor
0,meta-llama/Llama-2-7b-chat-hf,7574.67,22.89,0.0,8192-512,1,8193-512,1,sym_int4,N/A,3.21,12.5234375,N/A
# 4k input with split tensor
0,meta-llama/Llama-2-7b-chat-hf,2254.81,17.74,0.0,4096-512,1,4097-512,1,sym_int4,N/A,3.17,8.384765625,N/A
# 4k input without split tensor
0,meta-llama/Llama-2-7b-chat-hf,2217.67,17.71,0.0,4096-512,1,4097-512,1,sym_int4,N/A,3.09,8.361328125,N/A
```

The results shows, the split tensor optimization just affect the first token latency and do not affect rest token latency. Besides, the output of llama2-7b is abnormal when input length > 4096 (abnormal means repeat, garbled code and unknown code), yet the output results with and without split tensor are the same.

**4k input results:**
```shell
As the truth came into focus, the man realized that he had been given a second chance at life. He knew that his journey was far from his not alone in the city where he had lived his entire life. The deeper he delved into a stunning picture of a world that had been lost for centuries, a world where he had been hidden in perfect harmony. They discovered artifacts and documents that revealed the truth about their own talents of the city's secrets of the man's own had been hidden in the city's own had been lost in the city's of the man's own had been lost in the city's of the man's of the city's lost in the man's lost in the city's of the man's lost in the city's of the man's lost in the city's of the man's lost in the city's of the man's lost in the city's of the man's lost in the city's of the man's of the lost in the city's of the man's of the city's of the lost in the man's of the city's lost in the man's of the city's of the man's of the lost in the city's of the man's of the city's of the lost in the man's of the city's of the man's of the lost in the of the city's of the man's of the city's of the man's of the city's of the man's of the city's of the man's of the lost of the man's of the city's of
```

**6k input results:**
```shell
Љ
Љ
Љ
Љ
Љ
Љ
Љ
Љ
Љ
Љ.Љ.ЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉ.Љ.Љ
Љ
Љ
Љ
Љ
Љ
Љ
Љ.Љ
Љ.ЉЉЉ.Љ.Љ.Љ.Љ.Љ.Љ.ЉЉЉЉЉЉЉ.Љ.Љ.Љ.Љ.Љ.Љ.Љ.Љ.Љ.ЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉЉ.Љ
```

**8k input results:**
```shell
As he walked, the trees grew taller and closer together, until he was surrounded by a canopy of leaves that blocked out most of the light. He felt a sudden sense of unease, as if he was being watched. He tried to shake off the feeling<unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk><unk>
```

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 1]
```diff
@@ -639,11 +639,18 @@ def llama_attention_forward_4_31_original(
         attn_weights = None
     else:
         # otherwise, use native attention
-        attn_output, attn_weights = native_sdp(query_states, key_states, value_states,
+        if query_states.dtype == torch.float16 and query_states.shape[2] >= 6800:
```
Comment (hkvision) [inline python/llm/src/ipex_llm/transformers/models/llama.py:642]: also need to add the condition if don't return attn_weights?

**Comment by lalalapotter** [comment]
Noted that split tensor can also reduce the peak memory usage in 6.9k-8k input, when input >= 6900, without split tensor, report error:

```shell
  File "/home/arda/anaconda3/envs/cengguang-llm/lib/python3.9/site-packages/ipex_llm/transformers/models/llama.py", line 1320, in native_sdp
    attn_weights = torch.matmul(query.to(key.dtype),
RuntimeError: Allocation is out of device memory on current platform.
```

After using split tensor, it can run normally. Will test the memory reduction later.

**Comment by lalalapotter** [comment]
Tested with `bigdl_fp16_gpu `and `ipex_fp16_gpu `APIs, and get similar output results(with duplicates, garbled code) as mentioned above.

**Comment by lalalapotter** [comment]
Get normal results using [togethercomputer/Llama-2-7B-32K-Instruct](https://huggingface.co/togethercomputer/Llama-2-7B-32K-Instruct)

**8k input results:**
```shell
He walked on, his heart pounding in his chest. He knew that he was being foolish, that there was no one there. But the feeling persisted, as if he was being watched by an unseen force. He tried to shake off the feeling, but it lingered like a ghostly presence.
He walked on, his senses strained with the sounds of the forest. He could hear the rustling of leaves, the chirping of birds, and the distant sound of running water. But he could not shake off the feeling that he was being watched. He knew that he was alone in the forest, but he could not shake off the feeling that he was being followed.
[CHAPTER 7:   
The sun was setting over the horizon, casting long shadows across the dusty ground of the town square. The last rays of light streamed through the gaps between the buildings, illuminating the cobblestones and the people milling about. A group of children played a rough game of tag, their laughter filling the air.
In the center of the square stood a lone figure, a man with a tired face and weary eyes. He was tall and broad shouldered, but his posture was slumped and his head hung low. Despite the warmth of the day, he wore a thick woolen coat that
```

**Comment by lalalapotter** [comment]
Test 10k input results:
**perf test**
```shell
0,togethercomputer/Llama-2-7B-32K-Instruct,11220.54,22.23,0.0,10240-1024,1,10241-1024,1,sym_int4,N/A,6.63,10.912109375,N/A
```
**output result**
```shell
He took a step forward, his foot sinking into the soft earth. He felt a sudden jolt of recognition, as if he had been here before. But he couldn't remember when or why. He tried to shake off the feeling, but it lingered like a ghostly presence.
He walked deeper into the forest, the trees growing taller and closer together. The air grew thick with the scent of damp earth and decaying leaves. He could hear the distant sound of running water, the rush of a river or stream. He followed the sound, his heart pounding in his chest.
As he walked, the trees grew taller and closer together, until he was surrounded by a canopy of leaves that blocked out most of the light. He felt a sudden sense of unease, as if he was being watched. He tried to shake off the feeling, but it lingered like a ghostly presence.
He heard the sound of footsteps behind him, heavy and deliberate. He spun around, his hand already reaching for the hilt of his sword. But there was no one there. He was alone in the forest, surrounded by the silence of the trees.
He took a deep breath, trying to calm himself down. He knew that he was being foolish, that there was no one there. But the feeling persisted, as if he was being watched by an unseen force. He tried to shake off the feeling, but it lingered like a ghostly presence.
He walked on, his senses strained with the sounds of the forest. He could hear the rustling of leaves, the chirping of birds, and the distant sound of running water. But he could not shake off the feeling that he was being watched. He knew that he was alone in the forest, but he could not shake off the feeling that he was being followed
```

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 2]
```diff
@@ -639,11 +639,19 @@ def llama_attention_forward_4_31_original(
         attn_weights = None
     else:
         # otherwise, use native attention
-        attn_output, attn_weights = native_sdp(query_states, key_states, value_states,
-                                               attention_mask,
-                                               bsz, q_len, kv_seq_len,
-                                               self.head_dim, self.num_heads)
-
+        if not output_attentions and query_states.dtype == torch.float16 and \
+                query_states.shape[2] >= 6800:
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/llama.py:643]: refactor the check into a utility function

Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py]: Sure, fixed.

File: python/llm/src/ipex_llm/transformers/models/llama.py (hunk 1 above)
Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py]: fixed.

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 3]
```diff
@@ -214,6 +214,15 @@ def should_use_fast_rope(self, query_states, position_ids):
     return use_fuse_rope
 
 
+def should_split_tensor(query_states, output_attentions):
```
[the remark this answers is not in the document]
Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py]: fixed.

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 4]
```diff
@@ -1318,33 +1328,65 @@ def llama_attention_forward_4_36_original(
...
+        if attention_mask is not None:
+            attn_mask_size = (bsz, 1, q_len, kv_seq_len)
+            if attention_mask.size() != attn_mask_size:
+                invalidInputError(False,
+                                  f"Attention mask should be of size {attn_mask_size}, "
+                                  f"but is {attention_mask.size()}")
+            attn_weights_sp = attn_weights_sp + attention_mask
+        attn_weights_sp = nn.functional.softmax(attn_weights_sp, dim=-1)
+        attn_weights_sp = torch.matmul(attn_weights_sp, v)
+        attn_weights.append(attn_weights_sp)
```
Comment (hkvision) [inline python/llm/src/ipex_llm/transformers/models/llama.py:1387]: here should be attn_outputs_sp?

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 5]
```diff
@@ -1318,33 +1328,65 @@ def llama_attention_forward_4_36_original(
...
+        attn_output = torch.matmul(attn_weights, value)
+        return attn_output, attn_weights
+
+
+def native_sdp_split_qkv_tensor(query, key, value, attention_mask,
+                                bsz, q_len, kv_seq_len, head_dim):
+    query_sp = torch.split(query.to(key.dtype), 16, dim=1)
+    key_sp = torch.split(key.transpose(2, 3), 16, dim=1)
+    value_sp = torch.split(value, 16, dim=1)
+    attn_weights = []
```
Comment (hkvision) [inline python/llm/src/ipex_llm/transformers/models/llama.py:1369]: should be attn_outputs?

File: python/llm/src/ipex_llm/transformers/models/llama.py [hunk 6]
```diff
@@ -1318,33 +1328,65 @@ def llama_attention_forward_4_36_original(
...
+            # upcast attention to fp32
+            attn_weights = nn.functional.softmax(attn_weights, dim=-1,
+                                                 dtype=torch.float32).to(value.dtype)
+        attn_output = torch.matmul(attn_weights, value)
+        return attn_output, attn_weights
+
+
+def native_sdp_split_qkv_tensor(query, key, value, attention_mask,
+                                bsz, q_len, kv_seq_len, head_dim):
+    query_sp = torch.split(query.to(key.dtype), 16, dim=1)
```
Comment (hkvision) [inline python/llm/src/ipex_llm/transformers/models/llama.py:1366]: does sp mean split? Can we change for a better name to understand?

File: python/llm/src/ipex_llm/transformers/models/llama.py (hunk 4 above)
Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py:1387]: fixed.

File: python/llm/src/ipex_llm/transformers/models/llama.py (hunk 5 above)
Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py:1369]: fixed.

File: python/llm/src/ipex_llm/transformers/models/llama.py (hunk 6 above)
Comment (lalalapotter) [reply python/llm/src/ipex_llm/transformers/models/llama.py:1366]: fixed.

**Review by hkvision — APPROVED** [review]

**Pull request merged**